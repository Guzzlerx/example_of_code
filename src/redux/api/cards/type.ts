import { IMyCardProductItem } from 'routes/Main/Cards/CardProducts/type';
import { ICardStatusSwitcherProps } from 'routes/Main/Cards/MyCards/MyCardDetails/MyCardDetailsTabGroup/Tabs/ManageTab/Parts/CardStatusSwitcher/type';
import { IMyCardItem } from 'routes/Main/Cards/MyCards/type';
import {
  CARD_NAMES_ENUM,
  CARD_STATUSES_ENUM,
  CARD_TYPES_ENUM,
  CURRENCY_CODES_ENUM,
  PAYMENT_SYSTEMS_ENUM,
} from 'constants/cards/cards';

export interface IMyCardsList {
  cards: IMyCardItem[];
}

export interface IMyCardDetails {
  creditLimit: string;
  principalDebt: string;
  terminationDate: string;
  accountId: string;
  accountNumber: string;
  balance: string;
  cardId: string;
  cardNumber: string;
  clientId: string;
  currencyCode: CURRENCY_CODES_ENUM;
  expirationDate: string;
  holderName: string;
  paymentSystem: PAYMENT_SYSTEMS_ENUM;
  productName: CARD_NAMES_ENUM;
  status: CARD_STATUSES_ENUM;
  transactionLimit: string;
}

export interface ICardProductsList {
  cardProducts: IMyCardProductItem[];
}

export type TUpdateCardStatusRequest = Omit<ICardStatusSwitcherProps, 'refetchHandler'>;
export interface INewCardRequest {
  idProduct?: string;
  holderName?: string;
  cardType?: CARD_TYPES_ENUM;
}
