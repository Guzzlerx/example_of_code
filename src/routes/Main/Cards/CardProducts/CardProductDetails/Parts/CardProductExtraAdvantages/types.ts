import { ICardConfigsExtraAdvantages } from 'constants/cards/types';

export interface ICardProductExtraAdvantagesProps {
  data: ICardConfigsExtraAdvantages[];
}
