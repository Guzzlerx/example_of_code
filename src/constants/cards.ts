export enum CARD_TYPES_ENUM {
  DEBIT = 'DEBIT',
  CREDIT = 'CREDIT',
  DEPOSIT = 'DEPOSIT',
}

export enum CARD_STATUSES_ENUM {
  ACTIVE = 'ACTIVE',
  BLOCKED = 'BLOCKED',
  EXPIRED = 'EXPIRED',
}

export enum PAYMENT_SYSTEMS_ENUM {
  VISA = 'VISA',
  MASTERCARD = 'MASTERCARD',
}

export enum CARD_NAMES_ENUM {
  UNIVERSAL_CARD = 'universal_card',
  SMART_PAYOUT = 'smart_payout',
  PURCHASE_PLUS = 'purchase_plus',
  PERSONAL_LOAN = 'personal_loan',
  INFINITY_CLASSIC = 'infinity_classic',
  INFINITY_GOLD = 'infinity_gold',
  INFINITY_PREMIUM = 'infinity_premium',
}

export enum CURRENCY_CODES_ENUM {
  EUR = 'EUR',
  USD = 'USD',
  KZT = 'KZT',
  GBP = 'GBP',
  CHF = 'CHF',
  PLN = 'PLN',
  RUB = 'RUB',
}

export const MAX_CARDS_TO_SHOW = 3;

export enum CARD_SIZE_ENUM {
  LARGE = 'large',
  MEDIUM = 'medium',
  SMALL = 'small',
  EXTRA_SMALL = 'extraSmall',
}
