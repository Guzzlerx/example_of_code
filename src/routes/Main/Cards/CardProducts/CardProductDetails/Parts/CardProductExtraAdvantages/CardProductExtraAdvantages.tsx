import { Trans, useTranslation } from 'react-i18next';
import { AIcon } from 'components/AIcon/AIcon';
import { ATitle } from 'components/ATitle/ATitle';
import {
  StyledAdvantagesContainer,
  StyledAdvantagesWrapper,
  StyledContainer,
  StyledText,
} from '../CardProductAdvantages/CardProductAdvantages.styles';
import { ICardProductExtraAdvantagesProps } from './types';

export const CardProductExtraAdvantages = ({ data }: ICardProductExtraAdvantagesProps) => {
  const { t } = useTranslation('cards');

  return (
    <StyledContainer>
      <ATitle size={20} weight={700} lineHeight={28} marginBottom={40}>
        {t('card_details.extra_advantages.title')}
      </ATitle>
      <StyledAdvantagesContainer marginBottom={72}>
        {data.map(({ title, icon, description }, index) => (
          <StyledAdvantagesWrapper width={180} key={index}>
            <AIcon iconTitle={t(title)} variant="cardAdvantages">
              {icon}
            </AIcon>
            <StyledText>
              <Trans>{t(description)}</Trans>
            </StyledText>
          </StyledAdvantagesWrapper>
        ))}
      </StyledAdvantagesContainer>
    </StyledContainer>
  );
};
