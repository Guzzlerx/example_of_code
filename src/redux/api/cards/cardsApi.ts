import { apiSlice } from 'redux/api/apiSlice';
import { ICardStatusSwitcherResponse } from 'routes/Main/Cards/MyCards/MyCardDetails/MyCardDetailsTabGroup/Tabs/ManageTab/Parts/CardStatusSwitcher/type';
import { CARD_TYPES_ENUM } from 'constants/cards/cards';
import {
  ICardProductsList,
  IMyCardDetails,
  IMyCardsList,
  INewCardRequest,
  TUpdateCardStatusRequest,
} from './type';

export const cardsApi = apiSlice.injectEndpoints({
  endpoints: builder => ({
    getCreditCards: builder.query<IMyCardsList, void>({
      query: () => '/credit-cards',
    }),
    getDebitCards: builder.query<IMyCardsList, void>({
      query: () => '/deposit-cards',
    }),
    getCardDetails: builder.query<IMyCardDetails, { id: string; type: CARD_TYPES_ENUM }>({
      query: ({ id, type }) => `/${type.toLowerCase()}-cards/${id}`,
    }),
    getCardProducts: builder.query<ICardProductsList, void>({
      query: () => '/cards-products',
    }),
    updateCardStatus: builder.mutation<ICardStatusSwitcherResponse, TUpdateCardStatusRequest>({
      query: ({ cardId, currentStatus, cardType }) => ({
        url: `${cardType.toLocaleLowerCase()}-cards/${cardId}`,
        method: 'PATCH',
        body: currentStatus,
      }),
    }),
    getNewCard: builder.mutation<void, INewCardRequest>({
      query: body => ({
        url: '/deposit-cards-orders/new',
        method: 'POST',
        body: body,
      }),
    }),
  }),
});

export const {
  useGetCreditCardsQuery,
  useGetDebitCardsQuery,
  useGetCardProductsQuery,
  useGetCardDetailsQuery,
  useUpdateCardStatusMutation,
  useGetNewCardMutation,
} = cardsApi;
