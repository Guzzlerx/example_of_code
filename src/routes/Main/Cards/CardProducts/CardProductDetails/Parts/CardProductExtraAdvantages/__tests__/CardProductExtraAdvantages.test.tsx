import { RenderResult, render, screen } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next';
import { ReactComponent as Graph } from 'assets/icons/cardProductBenefits/deposit_graph.svg';
import { ReactComponent as EarnUp } from 'assets/icons/cardProductBenefits/earn_up.svg';
import { ReactComponent as Currency } from 'assets/icons/cardProductBenefits/euro_currency.svg';
import { ReactComponent as GracePeriod30 } from 'assets/icons/cardProductBenefits/grace_period.svg';
import { ReactComponent as MileProgram } from 'assets/icons/cardProductBenefits/mile_program.svg';
import { ReactComponent as Notifications } from 'assets/icons/cardProductBenefits/notifications.svg';
import { ReactComponent as Overdraft } from 'assets/icons/cardProductBenefits/overdraft.svg';
import { ReactComponent as VirtualCard } from 'assets/icons/cardProductBenefits/virtual_card.svg';
import { CARD_NAMES_ENUM } from 'constants/cards/cards';
import i18n from 'utils/languages/i18nForTests';
import { CardProductExtraAdvantages } from '../CardProductExtraAdvantages';

describe('CardProductExtraAdvantages component', () => {
  let renderResult: RenderResult;

  const data = [
    {
      title: `card_details.extra_advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.first_advantage.image_title`,
      description: `card_details.extra_advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.first_advantage.description`,
      icon: <Currency />,
    },
    {
      title: `card_details.extra_advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.second_advantage.image_title`,
      description: `card_details.extra_advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.second_advantage.description`,
      icon: <MileProgram />,
    },
    {
      title: `card_details.extra_advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.third_advantage.image_title`,
      description: `card_details.extra_advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.third_advantage.description`,
      icon: <Graph />,
    },
    {
      title: `card_details.extra_advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.fourth_advantage.image_title`,
      description: `card_details.extra_advantages.${CARD_NAMES_ENUM.UNIVERSAL_CARD}.icons.fourth_advantage.description`,
      icon: <GracePeriod30 />,
    },
  ];

  const rerenderData = [
    {
      title: `card_details.extra_advantages.${CARD_NAMES_ENUM.PURCHASE_PLUS}.icons.first_advantage.image_title`,
      description: `card_details.extra_advantages.${CARD_NAMES_ENUM.PURCHASE_PLUS}.icons.first_advantage.description`,
      icon: <VirtualCard />,
    },
    {
      title: `card_details.extra_advantages.${CARD_NAMES_ENUM.PURCHASE_PLUS}.icons.second_advantage.image_title`,
      description: `card_details.extra_advantages.${CARD_NAMES_ENUM.PURCHASE_PLUS}.icons.second_advantage.description`,
      icon: <EarnUp />,
    },
    {
      title: `card_details.extra_advantages.${CARD_NAMES_ENUM.PURCHASE_PLUS}.icons.third_advantage.image_title`,
      description: `card_details.extra_advantages.${CARD_NAMES_ENUM.PURCHASE_PLUS}.icons.third_advantage.description`,
      icon: <Overdraft />,
    },
    {
      title: `card_details.extra_advantages.${CARD_NAMES_ENUM.PURCHASE_PLUS}.icons.fourth_advantage.image_title`,
      description: `card_details.extra_advantages.${CARD_NAMES_ENUM.PURCHASE_PLUS}.icons.fourth_advantage.description`,
      icon: <Notifications />,
    },
  ];

  beforeEach(() => {
    renderResult = render(
      <I18nextProvider i18n={i18n}>
        <CardProductExtraAdvantages data={data} />
      </I18nextProvider>,
    );
  });

  it('should render correctly', () => {
    expect(renderResult.container).toBeInTheDocument();
  });

  it('snapshot', () => {
    expect(renderResult.container).toMatchSnapshot();
  });

  it('render correct icon wrapper width', () => {
    const element = screen.getByTitle(/Grace period/i);

    expect(element).toBeInTheDocument();
    expect(element).toHaveStyle('width: 64px');
  });

  it('render with another props', () => {
    renderResult.rerender(
      <I18nextProvider i18n={i18n}>
        <CardProductExtraAdvantages data={rerenderData} />
      </I18nextProvider>,
    );

    const description = screen.getByText(/Get yourself a virtual card to provide/i);
    const boldText = screen.getByText(/Â£ 2,000/i);

    expect(description).toBeInTheDocument();
    expect(boldText).toHaveStyle('font-weight: 700');
  });
});
