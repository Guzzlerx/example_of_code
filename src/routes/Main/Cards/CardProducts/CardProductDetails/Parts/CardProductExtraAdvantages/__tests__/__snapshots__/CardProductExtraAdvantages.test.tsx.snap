// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardProductExtraAdvantages component snapshot 1`] = `
<div>
  <section
    class="sc-dkrFOg eUBOWt"
  >
    <h2
      class="sc-gswNZR eugAJR"
      size="20"
    >
      You want more? We've got some
    </h2>
    <div
      class="sc-hLBbgP dsXhKp"
    >
      <div
        class="sc-eDvSVe bnkvWH"
        width="180"
      >
        <div
          class="sc-bcXHqe hhnFHT"
          title="Currency"
        >
          <svg>
            euro_currency.svg
          </svg>
        </div>
        <p
          class="sc-jSUZER gPgFBy"
        >
          An additional currency for your less spending and more gaining
        </p>
      </div>
      <div
        class="sc-eDvSVe bnkvWH"
        width="180"
      >
        <div
          class="sc-bcXHqe hhnFHT"
          title="Mile program"
        >
          <svg>
            mile_program.svg
          </svg>
        </div>
        <p
          class="sc-jSUZER gPgFBy"
        >
          Be part of our mile program. No queues and special benefits at the airport
        </p>
      </div>
      <div
        class="sc-eDvSVe bnkvWH"
        width="180"
      >
        <div
          class="sc-bcXHqe hhnFHT"
          title="Deposit graph"
        >
          <svg>
            deposit_graph.svg
          </svg>
        </div>
        <p
          class="sc-jSUZER gPgFBy"
        >
          Get timeless deposit up to 3% on your account balance
        </p>
      </div>
      <div
        class="sc-eDvSVe bnkvWH"
        width="180"
      >
        <div
          class="sc-bcXHqe hhnFHT"
          title="Grace period"
        >
          <svg>
            grace_period.svg
          </svg>
        </div>
        <p
          class="sc-jSUZER gPgFBy"
        >
          30 days of grace period. Itâ€™s good to can afford what you want
        </p>
      </div>
    </div>
  </section>
</div>
`;
